src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -654,21 +664,30 @@ public void ProduceContinue()
                 }
             }
 
-            if (_keepAlive == false && hasConnection == false && HttpVersion == "HTTP/1.1")
+            if (_keepAlive == false && _responseHeaders.HasConnection == false && _httpVersion == HttpVersionType.Http1_1)
             {
-                writer.Write("Connection: close\r\n\r\n");
+                _responseHeaders.SetRawConnection("close", _bytesConnectionClose);
             }
-            else if (_keepAlive && hasConnection == false && HttpVersion == "HTTP/1.0")
+            else if (_keepAlive && _responseHeaders.HasConnection == false && _httpVersion == HttpVersionType.Http1_0)
+            {
+                _responseHeaders.SetRawConnection("keep-alive", _bytesConnectionKeepAlive);
+            }
+
+            count += end.CopyFrom(_httpVersion == HttpVersionType.Http1_1 ? _bytesHttpVersion1_1 : _bytesHttpVersion1_0);
+            count += end.CopyFrom(statusBytes);
+            count += _responseHeaders.CopyTo(ref end);
+            count += end.CopyFrom(_bytesEndHeaders, 0, _bytesEndHeaders.Length);
+
+            SocketOutput.ProducingComplete(end, count);
+
+            if (immediate)
             {
-                writer.Write("Connection: keep-alive\r\n\r\n");
+                return SocketOutput.WriteAsync(default(ArraySegment<byte>), immediate: true);
```

Comment:
Flush doesn't exist on `ISocketOutput` and `ProducingComplete` stipulates "A write operation isn't guaranteed to be scheduled unless Write(Async) is called afterwards"